### PR TITLE
Add working frame rate control for all sensors to STM32

### PR DESCRIPTION
**Pull request by kwagyeman**
You just need to do set_framerate(fps) and the camera will run at a frame rate that's below or equal to the number passed. If you pass 0 then the camera goes back to the maximum fps. The code does a fairly good job of going exactly at the fps you want.

Only an implementation is available for the STM32. The code is smart and doesn't do work after it decides to drop a frame. Processor time/MDMA time is not wasted moving an image that will be tossed. Power is still wasted however transferring the image from the DCMI hardware through DMA2 to SRAM. But, this should be minimal.

Finally, this code works in conjunction with lowering the sensor frame rate. So, for each sensor you can run it slower too and you will get the correct result. Tested this on the Portenta.

### CODE REVIEW

File: src/omv/ports/stm32/sensor.c [hunk 1]
```diff
@@ -50,6 +50,8 @@ static DCMI_HandleTypeDef DCMIHandle = {.Instance = DCMI};
 static MDMA_HandleTypeDef DCMI_MDMA_Handle0 = {.Instance = MDMA_Channel0};
 static MDMA_HandleTypeDef DCMI_MDMA_Handle1 = {.Instance = MDMA_Channel1};
 #endif
+static bool DCMI_DMAConvCpltUser_first_line = false;
+static bool DCMI_DMAConvCpltUser_drop_frame = false;
```
Comment (iabdalkader) [inline src/omv/ports/stm32/sensor.c:54]:
```suggestion
static bool first_line = false;
static bool drop_frame = false;
```

File: src/omv/modules/py_sensor.c [hunk 2]
```diff
@@ -336,6 +336,10 @@ static mp_obj_t py_sensor_set_windowing(uint n_args, const mp_obj_t *args)
 
 static mp_obj_t py_sensor_get_windowing()
 {
+    if (sensor.framesize == FRAMESIZE_INVALID) {
+        mp_raise_msg(&mp_type_ValueError, MP_ERROR_TEXT("Frame size not set yet!"));
+    }
+
```
Comment (iabdalkader) [inline src/omv/modules/py_sensor.c:342]:
```suggestion
```
Doesn't look related, if not please send in another PR.

File: src/omv/common/sensor.h [hunk 3]
```diff
@@ -175,6 +175,8 @@ typedef struct _sensor {
     pixformat_t pixformat;      // Pixel format
     framesize_t framesize;      // Frame size
     int framerate;              // Frame rate
+    uint32_t last_frame_ms;     // Last sampled frame timestamp in milliseconds.
+    bool last_frame_ms_valid;   // Last sampled frame timestamp in milliseconds valid.
```
Comment (iabdalkader) [inline src/omv/common/sensor.h:179]:
```suggestion
```
Why not check if last_frame_ms is 0 ?

**Review by iabdalkader — CHANGES_REQUESTED** [review]

File: src/omv/common/sensor.h (hunk 3 above)
Comment (kwagyeman) [reply src/omv/common/sensor.h:179]: Wanted to make it as robust as possible. 0 is a valid time.

File: src/omv/ports/stm32/sensor.c (hunk 1 above)
Comment (kwagyeman) [reply src/omv/ports/stm32/sensor.c:54]: Done

File: src/omv/modules/py_sensor.c (hunk 2 above)
Comment (kwagyeman) [reply src/omv/modules/py_sensor.c:342]: Moved to other PR.

**Comment by kwagyeman** [comment]
Updated
